### PR TITLE
Use versioned types in kubectl describer

### PR DESCRIPTION
**Pull request by yue9944882**
/sig cli
/kind cleanup

**What this PR does / why we need it**:

this pull would be the last blocker to prune internal clientset 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/history.go [hunk 1]
```diff
@@ -168,12 +166,8 @@ func (h *DeploymentHistoryViewer) ViewHistory(namespace, name string, revision i
 
 func printTemplate(template *corev1.PodTemplateSpec) (string, error) {
 	buf := bytes.NewBuffer([]byte{})
-	internalTemplate := &api.PodTemplateSpec{}
```
Comment (yue9944882) [inline pkg/kubectl/history.go]: another explicit conversion gone

File: pkg/kubectl/cmd/convert/register.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+var Codecs = serializer.NewCodecFactory(Scheme)
+var ParameterCodec = runtime.NewParameterCodec(Scheme)
+
+func init() {
+	v1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+	Install(Scheme)
+}
+
+// Install registers the API group and adds types to a scheme
+func Install(scheme *runtime.Scheme) {
```
Comment (yue9944882) [inline pkg/kubectl/cmd/convert/register.go]: since the last dependency for internal client is gone in kubelet, the legacyscheme has to be initiated manually so that convert command can work correctly. (so i guess convert is also the last dependency for legacy scheme in kubectl too)

Comment (yue9944882) [reply pkg/kubectl/cmd/convert/register.go]: question: can convert run w/o legacy scheme? i suppose it's impossible (for now)

**Comment by Bronwyn645** [comment]
Please stop sending me e-mails

    On Tuesday, 13 November 2018, 1:35:44 PM AWST, k8s-ci-robot <notifications@github.com> wrote:  
 
 
@yue9944882: The following tests failed, say /retest to rerun them all:

| Test name | Commit | Details | Rerun command |
| pull-kubernetes-e2e-kops-aws | 0753956 | link | /test pull-kubernetes-e2e-kops-aws |
| pull-kubernetes-e2e-gce-100-performance | 0753956 | link | /test pull-kubernetes-e2e-gce-100-performance |
| pull-kubernetes-kubemark-e2e-gce-big | 0753956 | link | /test pull-kubernetes-kubemark-e2e-gce-big |
| pull-kubernetes-e2e-gce | 0753956 | link | /test pull-kubernetes-e2e-gce |
| pull-kubernetes-verify | e90ea10 | link | /test pull-kubernetes-verify |
| pull-kubernetes-integration | 5fdca5b | link | /test pull-kubernetes-integration |


Full PR test history. Your PR dashboard. Please help us cut down on flakes by linking to an open issue when you hit one in your PR.
 
Instructions for interacting with me using PR comments are available here. If you have questions or suggestions related to my behavior, please file an issue against the kubernetes/test-infra repository. I understand the commands that are listed here.

**Comment by yue9944882** [comment]
@Bronwyn645 plz click `Unsubscribe` in the page and u wont receive notifications no more.

**Comment by deads2k** [comment]
This is in keeping with the overall goals of both kubectl and client-go

/assign @pwittrock @soltysh

File: pkg/kubectl/cmd/convert/register.go (hunk 2 above)
Comment (yue9944882) [reply pkg/kubectl/cmd/convert/register.go]: > (so i guess convert is also the last dependency for legacy scheme in kubectl too)

no it isnt. we are printing internal types in get command via legacy scheme.

**Comment by yue9944882** [comment]
additionally, i tried to switch printers too [in this thread](https://github.com/kubernetes/kubernetes/pull/70861) but ended up finding out that it's also referenced by the inner printers of all REST storages. for printers, seems that we have to make a copy printer for external and keep the internal, which is a bit tougher than describer.

**Comment by soltysh** [comment]
> additionally, i tried to switch printers too [in this thread](https://github.com/kubernetes/kubernetes/pull/70861) but ended up finding out that it's also referenced by the inner printers of all REST storages. for printers, seems that we have to make a copy printer for external and keep the internal, which is a bit tougher than describer.

Yes, that's what we've been talking about during sig-cli meetings.

/cc @seans3

File: hack/.golint_failures [hunk 3]
```diff
@@ -250,6 +250,7 @@ pkg/master
 pkg/master/controller/crdregistration
 pkg/master/tunneler
 pkg/printers
+pkg/printers/externalversion
```
Comment (juanvallejo) [inline hack/.golint_failures]: An externalversion package for printers should go under pkg/kubectl

File: pkg/api/v1/events/sorted_event_list.go [hunk 4]
```diff
@@ -17,11 +17,11 @@ limitations under the License.
 package events
 
 import (
-	api "k8s.io/kubernetes/pkg/apis/core"
+	corev1 "k8s.io/api/core/v1"
```
Comment (juanvallejo) [inline pkg/api/v1/events/sorted_event_list.go]: Make a copy of this package under pkg/kubectl/util

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments

File: hack/.golint_failures (hunk 3 above)
Comment (soltysh) [reply hack/.golint_failures]: Yes, put all that under `pkg/kubectl/describe/`

File: pkg/kubectl/cmd/convert/import_known_versions.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package convert
+
+// These imports are the API groups the client will support.
+// TODO: Remove these manual install once we don't need legacy scheme in kubectl
+import (
```
Comment (soltysh) [inline pkg/kubectl/cmd/convert/import_known_versions.go]: Why is this needed?

File: hack/.golint_failures (hunk 3 above)
Comment (soltysh) [reply hack/.golint_failures]: While doing so drop the `versioned` sub-directory from there and just put everything in `pkg/kubectl/describe`.

File: pkg/kubectl/describe/versioned/describe.go [hunk 6]
```diff
@@ -23,7 +23,7 @@ import (
 	"k8s.io/cli-runtime/pkg/genericclioptions"
 	"k8s.io/kubernetes/pkg/kubectl/describe"
 	"k8s.io/kubernetes/pkg/printers"
-	printersinternal "k8s.io/kubernetes/pkg/printers/internalversion"
+	printersexternal "k8s.io/kubernetes/pkg/printers/externalversion"
```
Comment (juanvallejo) [inline pkg/kubectl/describe/versioned/describe.go]: Move this external printers package to `pkg/kubectl/printers`

Comment (juanvallejo) [reply pkg/kubectl/describe/versioned/describe.go]: Also, merge this file (`pkg/kubectl/describe/versioned/describe.go`) with `pkg/kubectl/describe/describe.go` and delete this `versioned` directory, we don't need it anymore

File: pkg/printers/externalversion/describe.go [hunk 7]
```diff
@@ -48,31 +58,21 @@ import (
...
-	"k8s.io/kubernetes/pkg/apis/core/helper"
-	"k8s.io/kubernetes/pkg/apis/core/helper/qos"
-	"k8s.io/kubernetes/pkg/apis/extensions"
-	"k8s.io/kubernetes/pkg/apis/networking"
-	"k8s.io/kubernetes/pkg/apis/policy"
-	"k8s.io/kubernetes/pkg/apis/rbac"
-	rbacv1helpers "k8s.io/kubernetes/pkg/apis/rbac/v1"
-	"k8s.io/kubernetes/pkg/apis/scheduling"
-	"k8s.io/kubernetes/pkg/apis/storage"
+	"k8s.io/kubernetes/pkg/api/v1/events"
```
Comment (seans3) [inline pkg/printers/externalversion/describe.go]: If this file is incorporated into `kubectl` as @juanvallejo suggests, the following seven imports are all problematic. These imports will increase the number of internal dependencies within `kubectl`, and make it more difficult to achieve `kubectl` independence.

File: pkg/printers/externalversion/describe.go [hunk 8]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package internalversion
+package externalversion
```
[the remark this answers is not in the document]
Comment (seans3) [reply pkg/printers/externalversion/describe.go]: @juanvallejo This PR is a great start. But since this file still has numerous dependencies on internal types, I'm not sure it is a good idea to move this file into `pkg/kubectl` yet. Moving this file into `pkg/kubectl` will make it more difficult to achieve `kubectl` independence. I suggest we do NOT move this file into `pkg/kubectl` until we've cleaned this file up more by removing the remaining internal dependencies.

File: pkg/kubectl/cmd/convert/BUILD [hunk 9]
```diff
@@ -2,12 +2,30 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
...
-    srcs = ["convert.go"],
+    srcs = [
+        "convert.go",
+        "import_known_versions.go",
+    ],
     importpath = "k8s.io/kubernetes/pkg/kubectl/cmd/convert",
     visibility = ["//visibility:public"],
     deps = [
         "//pkg/api/legacyscheme:go_default_library",
+        "//pkg/apis/apps/install:go_default_library",
```
Comment (seans3) [inline pkg/kubectl/cmd/convert/BUILD]: Agree with @soltysh. Not sure why we're adding these dependencies on internal types here. This is problematic. BTW, we are trying to deprecate `convert` or move it into a plugin.

File: pkg/kubectl/history.go [hunk 10]
```diff
@@ -168,12 +166,8 @@ func (h *DeploymentHistoryViewer) ViewHistory(namespace, name string, revision i
...
-	if err := apiv1.Convert_v1_PodTemplateSpec_To_core_PodTemplateSpec(template, internalTemplate, nil); err != nil {
-		return "", fmt.Errorf("failed to convert podtemplate, %v", err)
-	}
-	w := printersinternal.NewPrefixWriter(buf)
-	printersinternal.DescribePodTemplate(internalTemplate, w)
+	w := printersexternal.NewPrefixWriter(buf)
+	printersexternal.DescribePodTemplate(template, w)
```
Comment (seans3) [inline pkg/kubectl/history.go]: Very nice.

File: pkg/printers/externalversion/BUILD [hunk 11]
```diff
@@ -0,0 +1,90 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
+
+go_library(
+    name = "go_default_library",
+    srcs = ["describe.go"],
+    importpath = "k8s.io/kubernetes/pkg/printers/externalversion",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//pkg/api/v1/events:go_default_library",
```
Comment (seans3) [inline pkg/printers/externalversion/BUILD]: These thirteen dependencies are all problematic if this is moved into `pkg/kubectl`. We are currently at twenty-two remaining internal dependencies before `kubectl` independence. Moving this file into `pkg/kubectl` will dramatically increase this number.

**Review by seans3 — CHANGES_REQUESTED** [review]
This is a great PR, which is a significant help in moving `kubectl` out of core, and into its own repository. However, I do not agree with @juanvallejo that we should move code from `pkg/printers` into `pkg/kubectl` yet. Moving code that still contains numerous dependencies on internal types will only increase our load in this independence effort. I think we should keep `pkg/printers/externalversion/describe.go` in its current location.

File: hack/.golint_failures (hunk 3 above)
Comment (seans3) [reply hack/.golint_failures]: I disagree. Moving this code into `pkg/kubectl` will increase the current internal, core dependencies. I believe we should continue working on `pkg/printers/externalversion` until it is free of internal dependencies--then move it into `pkg/kubectl`.

File: pkg/printers/externalversion/describe.go [hunk 12]
```diff
@@ -138,43 +149,42 @@ func describerMap(clientConfig *rest.Config) (map[schema.GroupKind]printers.Desc
...
-		certificates.Kind("CertificateSigningRequest"): &CertificateSigningRequestDescriber{c},
-		storage.Kind("StorageClass"):                   &StorageClassDescriber{c},
-		policy.Kind("PodDisruptionBudget"):             &PodDisruptionBudgetDescriber{c},
-		rbac.Kind("Role"):                              &RoleDescriber{externalclient},
-		rbac.Kind("ClusterRole"):                       &ClusterRoleDescriber{externalclient},
-		rbac.Kind("RoleBinding"):                       &RoleBindingDescriber{externalclient},
-		rbac.Kind("ClusterRoleBinding"):                &ClusterRoleBindingDescriber{externalclient},
-		networking.Kind("NetworkPolicy"):               &NetworkPolicyDescriber{c},
-		scheduling.Kind("PriorityClass"):               &PriorityClassDescriber{c},
+		{Group: corev1.GroupName, Kind: "Pod"}:                                    &PodDescriber{c},
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply pkg/printers/externalversion/describe.go]: note that this is GroupKind, initially i used:

`corev1.SchemeGroupVersion.GroupVersion().WithKind("Pod").GroupKind()`

but this looks odd, isnt it? why is a straight struct initialization not prefered?

File: pkg/kubectl/cmd/convert/import_known_versions.go (hunk 5 above)
Comment (yue9944882) [reply pkg/kubectl/cmd/convert/import_known_versions.go]: this is the tricky part: as the last dependency for internal client is gone, the legacyscheme which is used by convert would be empty unless we manually install internal types. and thats the purpose of this file. if u notice, we are doing the same in `cmd/kube-controller-manager` for the exact same reason. w/o this file, the convert wont be working b/c legacyscheme doesn't know how to deal w/ conversions.

File: pkg/kubectl/cmd/convert/BUILD (hunk 9 above)
Comment (yue9944882) [reply pkg/kubectl/cmd/convert/BUILD]: > move it into a plugin

this one sounds better to me

**Comment by seans3** [comment]
The following two "describe"-related test instances are failing in e2e tests with this PR:

```
[sig-cli] Kubectl client [k8s.io] Kubectl describe should check if kubectl describe prints relevant information for rc and pods [Conformance] 1m36s

test/e2e/framework/framework.go:699
Nov 14 08:11:49.756: failed to find 1 desired in Replicas:     1 current / 824639146376 desired
test/e2e/kubectl/kubectl.go:1915

test/e2e/framework/framework.go:699
Nov 14 08:10:10.519: failed to find 1 desired in Replicas:     1 current / 824642171896 desired
test/e2e/kubectl/kubectl.go:1915
```

File: pkg/kubectl/describe/versioned/describe.go [hunk 13]
```diff
@@ -17,32 +17,4440 @@ limitations under the License.
...
+	// labelNodeRolePrefix is a label prefix for node roles
+	// It's copied over to here until it's merged in core: https://github.com/kubernetes/kubernetes/pull/39112
+	labelNodeRolePrefix = "node-role.kubernetes.io/"
+
+	// nodeLabelRole specifies the role of a node
+	nodeLabelRole = "kubernetes.io/role"
 )
 
 // DescriberFn gives a way to easily override the function for unit testing if needed
 var DescriberFn describe.DescriberFunc = Describer
```
Comment (soltysh) [inline pkg/kubectl/describe/versioned/describe.go]: Move this to `pkg/kubectl/describe` instead of here. It reads weird and might cause cycles.

File: pkg/kubectl/describe/describe.go [hunk 14]
```diff
@@ -17,10 +17,43 @@ limitations under the License.
 package describe
```
Comment (soltysh) [inline pkg/kubectl/describe/describe.go]: Rename this file to `interfaces.go` it'll work similar to what we have in polymorphichelpers.

File: pkg/kubectl/util/resource/resource.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"fmt"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/sets"
```
Comment (soltysh) [inline pkg/kubectl/util/resource/resource.go]: Split built-in and k8s imports.

File: pkg/kubectl/util/ref/ref.go [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+var (
+	ErrNilObject  = errors.New("can't reference a nil object")
+	ErrNoSelfLink = errors.New("selfLink was empty, can't make reference")
+)
+
+// GetReference returns an ObjectReference which refers to the given
+// object, or an error if the object doesn't follow the conventions
+// that would allow this.
+// TODO: should take a meta.Interface see http://issue.k8s.io/7127
+func GetReference(scheme *runtime.Scheme, obj runtime.Object) (*corev1.ObjectReference, error) {
```
Comment (soltysh) [inline pkg/kubectl/util/ref/ref.go]: Don't copy something that is already available in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/reference/ref.go#L41. If need to update references to this.

**Review by soltysh — CHANGES_REQUESTED** [review]
A few more comments, also please drop `pkg/kubectl/util/event/OWNERS` file, it's not needed in kubectl sub-dirs. And fix the tests that Sean pointed out.

File: pkg/kubectl/describe/versioned/describe.go [hunk 17]
```diff
@@ -17,32 +17,4440 @@ limitations under the License.
...
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	externalclient "k8s.io/client-go/kubernetes"
+	coreclient "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	ref "k8s.io/client-go/tools/reference"
```
Comment (soltysh) [inline pkg/kubectl/describe/versioned/describe.go]: Drop alias, fix all occurrences.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/kubectl/describe/versioned/describe.go [hunk 18]
```diff
@@ -17,32 +17,4440 @@ limitations under the License.
...
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	externalclient "k8s.io/client-go/kubernetes"
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply pkg/kubectl/describe/versioned/describe.go]: omg 😱nice catch. (how i hope go compiler can catch this for me)

**Comment by seans3** [comment]
LGTM

Thanks for all the work on this PR, and thanks for the quick turn-around on all the feedback.

File: pkg/kubectl/describe/versioned/describe.go (hunk 13 above)
Comment (soltysh) [reply pkg/kubectl/describe/versioned/describe.go]: This comment isn't still addressed.

File: pkg/kubectl/describe/versioned/describe_test.go [hunk 19]
```diff
@@ -26,46 +26,45 @@ import (
...
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	storagev1 "k8s.io/api/storage/v1"
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
 	versionedfake "k8s.io/client-go/kubernetes/fake"
```
Comment (soltysh) [inline pkg/kubectl/describe/versioned/describe_test.go]: Duplicate imports.

**Review by soltysh — APPROVED** [review]
Two more nits you can fix in a followup.
/lgtm
/approve

**Comment by yue9944882** [comment]
ref fixer: #71553
